File: Algorithms/src/1622. Fancy Sequence/Code/solution.js
var Fancy = function () {
  this.MOD = 1000000007n;
  this.seq = [];
  this.mul = 1n;
  this.add = 0n;
};

Fancy.prototype.modPow = function (a, b) {
  let res = 1n;
  a = BigInt(a);
  b = BigInt(b);

  while (b > 0n) {
    if (b & 1n) res = (res * a) % this.MOD;
    a = (a * a) % this.MOD;
    b >>= 1n;
  }
  return res;
};

Fancy.prototype.append = function (val) {
  val = BigInt(val);
  let inv = this.modPow(this.mul, this.MOD - 2n);
  let stored = (((val - this.add + this.MOD) % this.MOD) * inv) % this.MOD;
  this.seq.push(stored);
};

Fancy.prototype.addAll = function (inc) {
  this.add = (this.add + BigInt(inc)) % this.MOD;
};

Fancy.prototype.multAll = function (m) {
  m = BigInt(m);
  this.mul = (this.mul * m) % this.MOD;
  this.add = (this.add * m) % this.MOD;
};

Fancy.prototype.getIndex = function (idx) {
  if (idx >= this.seq.length) return -1;
  let val = (this.seq[idx] * this.mul + this.add) % this.MOD;
  return Number(val);
};

File: Algorithms/src/1622. Fancy Sequence/Code/solution.py
class Fancy:

    MOD = 10**9 + 7

    def __init__(self):
        self.seq = []
        self.mul = 1
        self.add = 0

    def mod_pow(self, a, b):
        res = 1
        while b:
            if b & 1:
                res = res * a % self.MOD
            a = a * a % self.MOD
            b >>= 1
        return res

    def append(self, val: int) -> None:
        inv = self.mod_pow(self.mul, self.MOD - 2)
        stored = ((val - self.add) % self.MOD) * inv % self.MOD
        self.seq.append(stored)

    def addAll(self, inc: int) -> None:
        self.add = (self.add + inc) % self.MOD

    def multAll(self, m: int) -> None:
        self.mul = self.mul * m % self.MOD
        self.add = self.add * m % self.MOD

    def getIndex(self, idx: int) -> int:
        if idx >= len(self.seq):
            return -1
        return (self.seq[idx] * self.mul + self.add) % self.MOD